File: 764-n-ary-tree-level-order-traversal/n-ary-tree-level-order-traversal.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     int numChildren;
 *     struct Node** children;
 * };
 */

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume
 * caller calls free().
 */
int** levelOrder(struct Node* root, int* returnSize, int** returnColumnSizes) {
    // Malloc necessary arrays
    int** result = malloc(1000 * sizeof(int*));
    int* columnSizes = malloc(1000 * sizeof(int));
    *returnColumnSizes = columnSizes;

    // Initialize *returnSize and handle empty tree edge case
    *returnSize = 0;
    if (!root)
        return result;

    /* In this solution we will use a queue to keep track of children
    The logic is to that in each iteration we will have the current level nodes
    in the queue, the when a node is treated (its val is gathered and all its
    children added to end of queue), we will we will move forward to the next
    node in the queue.
    */
    struct Node** queue = malloc(10001 * sizeof(struct Node*));
    int head = 0, tail = 0, i, j, temp;
    queue[tail++] = root;
    while (head < tail) {
        columnSizes[*returnSize] = tail - head;
        result[*returnSize] = malloc(columnSizes[*returnSize] * sizeof(int));
        temp = tail;
        for (i = head; i < temp; i++) {
            result[*returnSize][i - head] = queue[i]->val;
            for (j = 0; j < queue[i]->numChildren; j++) {
                queue[tail++] = queue[i]->children[j];
            }
        }
        head = i;
        (*returnSize)++;
    }
    return result;
}

File: 764-n-ary-tree-level-order-traversal/n-ary-tree-level-order-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: Optional[int] = None, children: Optional[List['Node']] = None):
        self.val = val
        self.children = children
"""


class Solution:
    def levelOrder(self, root: "Node") -> List[List[int]]:
        # Empty tree edge case
        if not root:
            return []

        # Initialize lists that will be used as well as the result list
        curr = [root]
        result = [[root.val]]

        # Loop through all nodes in the current level getting there values and there children
        while curr:
            temp = []
            next_level = []
            for node in curr:
                for child in node.children:
                    next_level.append(child)
                    temp.append(child.val)
            if temp:
                result.append(temp)
            curr = next_level
        return result
